File: Python/tailRecursion.py
# A tail recursive function
# to calculate factorial
def fact(n, a = 1):
 
    if (n == 1):
        return a
 
    return fact(n - 1, n * a)
 
# Driver program to test
#  above function
print(fact(5))
 
# This code is contributed
# by Smitha
# improved by Ujwal, ashish2021

File: CPP/tailRecursion.cpp
#include<iostream>
using namespace std;
 
// A tail recursive function to calculate factorial
unsigned factTR(unsigned int n, unsigned int a)
{
    if (n == 1)  return a;
 
    return factTR(n-1, n*a);
}
 
// A wrapper over factTR
unsigned int fact(unsigned int n)
{
   return factTR(n, 1);
}
 
// Driver program to test above function
int main()
{
    cout << fact(5);
    return 0;
}